File: src/main/java/io/dev/mutex/thesisinfomgmt/service/DegreeService.java
package io.dev.mutex.thesisinfomgmt.service;

import io.dev.mutex.thesisinfomgmt.common.PaginatedData;
import io.dev.mutex.thesisinfomgmt.dto.DegreeDTO;

public interface DegreeService {

  /**
   * Retrieves degrees
   *
   * @param page     the page of the data to be retrieved
   * @param pageSize the number of items to be retrieved
   * @param query    the query string
   * @return paginated degree data
   */
  PaginatedData<DegreeDTO> getDegrees(int page, int pageSize, String query);

  /**
   * Creates a new degree
   *
   * @param degree the details of the new degree
   * @return created degree
   */
  DegreeDTO createDegree(DegreeDTO degree);

  /**
   * Retrieves a degree by its id
   *
   * @param id the id of the degree to be retrieved
   * @return retrieved degree
   */
  DegreeDTO getDegree(long id);

  /**
   * Deletes a degree
   *
   * @param id the id of the degree to be deleted
   */
  void deleteDegree(long id);

  /**
   * Updates a degree
   *
   * @param id     the id of the degree to be updated
   * @param degree the updated degree details
   */
  void updateDegree(long id, DegreeDTO degree);
}

File: app/src/services/DegreeService.ts
import { PaginatedData } from "../types/Common";
import { Degree } from "../types/Degrees";
import RequestHelper from "../helpers/RequestHelper";
import qs from "qs";
import { RequestMethod } from "../common/constants";

class DegreeService {
  resourcePath = "degrees";

  async getDegrees(
    page: number,
    pageSize: number,
    query: string
  ): Promise<PaginatedData<Degree>> {
    const queryParams = {
      page: page,
      pageSize: pageSize,
      query: query,
    };
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}?${qs.stringify(queryParams)}`,
      "get degrees"
    );
  }

  async createDegree(degree: Degree): Promise<Degree> {
    return await RequestHelper.doRemoteCallAsync(
      this.resourcePath,
      "create degree",
      RequestMethod.POST,
      degree
    );
  }

  async updateDegree(degreeId: number, degree: Degree): Promise<void> {
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}/${degreeId}`,
      "update degree",
      RequestMethod.PUT,
      degree
    );
  }

  async deleteDegree(degreeId: number): Promise<void> {
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}/${degreeId}`,
      "delete degree",
      RequestMethod.DELETE
    );
  }
}

export default new DegreeService();
